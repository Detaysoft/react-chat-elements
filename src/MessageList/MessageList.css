.rce-container-mlist {
    position: relative;
    display: flex;
}

.rce-mlist {
    display: block;
    overflow: auto;
    position: relative;
    flex: 1;
}

.rce-mlist-down-button {
    position: absolute;
    right: 10px;
    bottom: 15px;
    width: 40px;
    height: 40px;
    background: #fff;
    box-shadow: 0 1px 1px 0 rgba(0,0,0,0.05), 0 2px 5px 0 rgba(0,0,0,0.1);
    border-radius: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    color: #333;
    cursor: pointer;
    transition: 200ms;
}

.rce-mlist-down-button:hover {
    opacity: 0.7;
}

.rce-mlist-down-button--badge {
    position: absolute;
    right: -5px;
    top: -5px;
    background: red;
    width: 20px;
    height: 20px;
    border-radius: 100%;
    font-size: 12px;
    display: flex;
    text-align: center;
    align-items: center;
    justify-content: center;
    color: #fff;
    font-weight: 700;
}
