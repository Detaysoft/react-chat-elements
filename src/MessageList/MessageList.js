import React, { Component } from 'react';
import './MessageList.css';

import MessageBox from '../MessageBox/MessageBox';

import FaChevronDown from 'react-icons/lib/fa/chevron-down';

const classNames = require('classnames');

export class MessageList extends Component {

    constructor(props) {
        super(props);
        this.state = {
            scrollBottom: 0,
            downButton: false,
        };
    }

    checkScroll() {
        var e = this.mlistRef;
        if (!e)
            return;

        if (this.props.toBottomHeight === '100%' || this.state.scrollBottom < this.props.toBottomHeight) {
            // scroll to bottom
            e.scrollTop = e.scrollHeight;
        } else {
            if (this.props.lockable === true) {
                e.scrollTop = e.scrollHeight - e.offsetHeight - this.state.scrollBottom;
            }
        }
    }

    componentWillReceiveProps() {
        if (!this.mlistRef)
            return;
        this.setState({
            scrollBottom: this.getBottom(this.mlistRef),
        }, this.checkScroll.bind(this));
    }

    getBottom(e) {
        return e.scrollHeight - e.scrollTop - e.offsetHeight;
    }

    onOpen(item, i, e) {
        if (this.props.onOpen instanceof Function)
            this.props.onOpen(item, i, e);
    }

    onDownload(item, i, e) {
        if (this.props.onDownload instanceof Function)
            this.props.onDownload(item, i, e);
    }

    onClick(item, i, e) {
        if (this.props.onClick instanceof Function)
            this.props.onClick(item, i, e);
    }

    onTitleClick(item, i, e) {
        if (this.props.onTitleClick instanceof Function)
            this.props.onTitleClick(item, i, e);
    }

    onForwardClick(item, i, e) {
        if (this.props.onForwardClick instanceof Function)
            this.props.onForwardClick(item, i, e);
    }

    loadRef(ref) {
        this.mlistRef = ref;
        if (this.props.cmpRef instanceof Function)
            this.props.cmpRef(ref);
    }

    onScroll(e) {
        var bottom = this.getBottom(e.target);
        this.state.scrollBottom = bottom;
        if (this.props.toBottomHeight === '100%' || bottom > this.props.toBottomHeight) {
            if (this.state.downButton !== true) {
                this.state.downButton = true;
                this.setState({
                    downButton: true,
                    scrollBottom: bottom,
                })
            }
        } else {
            if (this.state.downButton !== false) {
                this.state.downButton = false;
                this.setState({
                    downButton: false,
                    scrollBottom: bottom,
                })
            }
        }

        if (this.props.onScroll instanceof Function) {
            this.props.onScroll(e);
        }
    }

    toBottom(e) {
        if(!this.mlistRef)
            return;
        this.mlistRef.scrollTop = this.mlistRef.scrollHeight;
        if (this.props.onDownButtonClick instanceof Function) {
            this.props.onDownButtonClick(e);
        }
    }

    render() {
        return (
            <div
                className={classNames(['rce-container-mlist', this.props.className])}>
                <div
                    ref={this.loadRef.bind(this)}
                    onScroll={this.onScroll.bind(this)}
                    className='rce-mlist'>
                    {
                        this.props.dataSource.map((x, i) => (
                            <MessageBox
                                key={i}
                                {...x}
                                onOpen={this.props.onOpen && ((e) => this.onOpen(x, i, e))}
                                onDownload={this.props.onDownload && ((e) => this.onDownload(x, i, e))}
                                onTitleClick={this.props.onDownload && ((e) => this.onTitleClick(x, i, e))}
                                onForwardClick={this.props.onForwardClick && ((e) => this.onForwardClick(x, i, e))}
                                onClick={this.props.onClick && ((e) => this.onClick(x, i, e))}/>
                        ))
                    }
                </div>
                {
                    this.props.downButton === true &&
                    this.state.downButton &&
                    this.props.toBottomHeight !== '100%' &&
                    <div
                        className='rce-mlist-down-button'
                        onClick={this.toBottom.bind(this)}>
                        <FaChevronDown/>
                        {
                            this.props.downButtonBadge &&
                            <span
                                className='rce-mlist-down-button--badge'>
                                {this.props.downButtonBadge}
                            </span>
                        }
                    </div>
                }
            </div>
        );
    }
}

MessageList.defaultProps = {
    onClick: null,
    onTitleClick: null,
    onForwardClick: null,
    onDownButtonClick: null,
    onOpen: null,
    onDownload: null,
    dataSource: [],
    lockable: false,
    toBottomHeight: 300,
    downButton: true,
    downButtonBadge: null,
};

export default MessageList;
