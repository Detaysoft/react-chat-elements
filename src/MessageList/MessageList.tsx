import React, { FC, Key, useEffect, useRef, useState } from 'react'
import './MessageList.css'

import MessageBox from '../MessageBox/MessageBox'

import classNames from 'classnames'
import { FaChevronDown } from 'react-icons/fa'
import { IMessageListProps, MessageListEvent, MessageType } from '../type'

import { Virtuoso } from 'react-virtuoso'

const MessageList: FC<IMessageListProps<{}, {}>> = ({
  referance = null,
  lockable = false,
  toBottomHeight = 300,
  downButton,
  ...props
}) => {
  const [scrollBottom, setScrollBottom] = useState(0)
  const [_downButton, setDownButton] = useState(false)
  const prevProps = useRef(props)

  const checkScroll = () => {
    var e = referance
    if (!e || !e.current) return

    if (props.virtualListOptions) {
      setTimeout(() => {
        e.current.scrollToIndex({
          index: props.dataSource.length - 1,
          align: 'end',
        })
      }, 0)
      return
    }

    if (toBottomHeight === '100%' || (toBottomHeight && scrollBottom < toBottomHeight)) {
      e.current.scrollTop = e.current.scrollHeight // scroll to bottom
    } else {
      if (lockable === true) {
        e.current.scrollTop = e.current.scrollHeight - e.current.offsetHeight - scrollBottom
      }
    }
  }

  useEffect(() => {
    fitScrollToBottom()
  }, [])

  useEffect(() => {
    if (!referance) return

    if (prevProps.current.dataSource.length !== props.dataSource.length) {
      fitScrollToBottom()
    }

    prevProps.current = props
  }, [prevProps, props])

  const fitScrollToBottom = () => {
    setScrollBottom(getBottom(referance))
    checkScroll()
  }

  const getBottom = (e: any) => {
    if (e.current) return e.current.scrollHeight - e.current.scrollTop - e.current.offsetHeight
    return e.scrollHeight - e.scrollTop - e.offsetHeight
  }

  const onScroll = (e: React.UIEvent<HTMLElement>): void => {
    var bottom = getBottom(e.currentTarget)
    setScrollBottom(bottom)
    if (toBottomHeight === '100%' || (toBottomHeight && bottom > toBottomHeight)) {
      if (_downButton !== true) {
        setDownButton(true)
        setScrollBottom(bottom)
      }
    } else {
      if (_downButton !== false) {
        setDownButton(false)
        setScrollBottom(bottom)
      }
    }

    if (props.onScroll instanceof Function) {
      props.onScroll(e)
    }
  }

  const toBottom = (e: any) => {
    if (!referance) return

    if (props.virtualListOptions !== null) {
      referance.current.scrollToIndex({ index: props.dataSource.length - 1, behavior: 'smooth' })
      return
    }

    referance.current.scrollTop = referance.current.scrollHeight
    if (props.onDownButtonClick instanceof Function) {
      props.onDownButtonClick(e)
    }
  }

  return (
    <div className={classNames(['rce-container-mlist', props.className])} {...props.customProps}>
      {!!props.children && props.isShowChild && props.children}

      {props.virtualListOptions ? (
        <div className={classNames(['rce-container-mlist', props.className])} {...props.customProps}>
          <Virtuoso
            ref={referance}
            // scrollerRef={ref => (referance.current = ref)}
            // followOutput={true}
            onScroll={onScroll}
            className='rce-mlist'
            style={{ height: '100%' }}
            {...props.virtualListOptions}
            totalCount={props.dataSource.length}
            atBottomStateChange={bottom => {
              setDownButton(!bottom)
            }}
            itemContent={index => <MeessageContent index={index} item={props.dataSource[index]} props={props} />}
          />
        </div>
      ) : (
        <div ref={referance} onScroll={onScroll} className='rce-mlist'>
          {props.dataSource.map((item, i: number) => (
            <MeessageContent index={i} item={item} props={props} />
          ))}
        </div>
      )}

      {downButton === true && _downButton && toBottomHeight !== '100%' && (
        <div className='rce-mlist-down-button' onClick={toBottom}>
          <FaChevronDown />
          {props.downButtonBadge !== undefined ? (
            <span className='rce-mlist-down-button--badge'>{props.downButtonBadge.toString()}</span>
          ) : null}
        </div>
      )}
    </div>
  )
}

function MeessageContent({ props, item, index }: { props: any; item: MessageType; index: number }) {
  const onOpen: MessageListEvent = (item, index, event) => {
    if (props.onOpen instanceof Function) props.onOpen(item, index, event)
  }

  const onDownload: MessageListEvent = (item, index, event) => {
    if (props.onDownload instanceof Function) props.onDownload(item, index, event)
  }

  const onPhotoError: MessageListEvent = (item, index, event) => {
    if (props.onPhotoError instanceof Function) props.onPhotoError(item, index, event)
  }

  const onClick: MessageListEvent = (item, index, event) => {
    if (props.onClick instanceof Function) props.onClick(item, index, event)
  }

  const onTitleClick: MessageListEvent = (item, index, event) => {
    if (props.onTitleClick instanceof Function) props.onTitleClick(item, index, event)
  }

  const onForwardClick: MessageListEvent = (item, index, event) => {
    if (props.onForwardClick instanceof Function) props.onForwardClick(item, index, event)
  }

  const onReplyClick: MessageListEvent = (item, index, event) => {
    if (props.onReplyClick instanceof Function) props.onReplyClick(item, index, event)
  }

  const onReplyMessageClick: MessageListEvent = (item, index, event) => {
    if (props.onReplyMessageClick instanceof Function) props.onReplyMessageClick(item, index, event)
  }

  const onRemoveMessageClick: MessageListEvent = (item, index, event) => {
    if (props.onRemoveMessageClick instanceof Function) props.onRemoveMessageClick(item, index, event)
  }

  const onContextMenu: MessageListEvent = (item, index, event) => {
    if (props.onContextMenu instanceof Function) props.onContextMenu(item, index, event)
  }

  const onMessageFocused: MessageListEvent = (item, index, event) => {
    if (props.onMessageFocused instanceof Function) props.onMessageFocused(item, index, event)
  }

  const onMeetingMessageClick: MessageListEvent = (item, index, event) => {
    if (props.onMeetingMessageClick instanceof Function) props.onMeetingMessageClick(item, index, event)
  }

  const onMeetingMoreSelect: MessageListEvent = (item, i, e) => {
    if (props.onMeetingMoreSelect instanceof Function) props.onMeetingMoreSelect(item, i, e)
  }

  const onMeetingLinkClick: MessageListEvent = (item, i, e) => {
    if (props.onMeetingLinkClick instanceof Function) props.onMeetingLinkClick(item, i, e)
  }
  return (
    <MessageBox
      key={index as Key}
      {...(item as any)}
      // data={x}
      onOpen={props.onOpen && ((e: React.MouseEvent<HTMLElement>) => onOpen(item, index, e))}
      onPhotoError={props.onPhotoError && ((e: React.MouseEvent<HTMLElement>) => onPhotoError(item, index, e))}
      onDownload={props.onDownload && ((e: React.MouseEvent<HTMLElement>) => onDownload(item, index, e))}
      onTitleClick={props.onTitleClick && ((e: React.MouseEvent<HTMLElement>) => onTitleClick(item, index, e))}
      onForwardClick={props.onForwardClick && ((e: React.MouseEvent<HTMLElement>) => onForwardClick(item, index, e))}
      onReplyClick={props.onReplyClick && ((e: React.MouseEvent<HTMLElement>) => onReplyClick(item, index, e))}
      onReplyMessageClick={
        props.onReplyMessageClick && ((e: React.MouseEvent<HTMLElement>) => onReplyMessageClick(item, index, e))
      }
      onRemoveMessageClick={
        props.onRemoveMessageClick && ((e: React.MouseEvent<HTMLElement>) => onRemoveMessageClick(item, index, e))
      }
      onClick={props.onClick && ((e: React.MouseEvent<HTMLElement>) => onClick(item, index, e))}
      onContextMenu={props.onContextMenu && ((e: React.MouseEvent<HTMLElement>) => onContextMenu(item, index, e))}
      onMeetingMoreSelect={
        props.onMeetingMoreSelect && ((e: React.MouseEvent<HTMLElement>) => onMeetingMoreSelect(item, index, e))
      }
      onMessageFocused={props.onMessageFocused && ((e: React.MouseEvent<HTMLElement>) => onMessageFocused(item, index, e))}
      onMeetingMessageClick={
        props.onMeetingMessageClick && ((e: React.MouseEvent<HTMLElement>) => onMeetingMessageClick(item, index, e))
      }
      onMeetingTitleClick={props.onMeetingTitleClick}
      onMeetingVideoLinkClick={props.onMeetingVideoLinkClick}
      onMeetingLinkClick={props.onMeetingLinkClick && ((e: React.MouseEvent<HTMLElement>) => onMeetingLinkClick(item, index, e))}
      actionButtons={props.actionButtons}
      styles={props.messageBoxStyles}
      notchStyle={props.notchStyle}
    />
  )
}

export default MessageList
