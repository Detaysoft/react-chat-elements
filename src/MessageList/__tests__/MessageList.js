import React, { Component } from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MessageList from '../MessageList';

describe('MessageList component', () => {
  it('should render without issues', () => {
    const component = shallow(<MessageList />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });
});
