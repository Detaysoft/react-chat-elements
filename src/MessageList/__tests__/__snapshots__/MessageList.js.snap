// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList component should render without issues 1`] = `
<div
  className="rce-container-mlist"
>
  <div
    className="rce-mlist"
    onScroll={[Function]}
  />
</div>
`;
